authz/rbac_translator.go

```diff
@@ -0,0 +1,301 @@
+	return &v3matcherpb.StringMatcher{
+		MatchPattern: &v3matcherpb.StringMatcher_Exact{
+			Exact: value,
+		},
+	}
+}
+
+func getHeaderMatcher(key, value string) *v3routepb.HeaderMatcher {
+	if value == "*" {
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_PrefixMatch{PrefixMatch: ""},
+		}
+	} else if strings.HasSuffix(value, "*") {
+		prefix := strings.TrimSuffix(value, "*")
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_PrefixMatch{PrefixMatch: prefix},
+		}
+	} else if strings.HasPrefix(value, "*") {
+		suffix := strings.TrimPrefix(value, "*")
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_SuffixMatch{SuffixMatch: suffix},
+		}
+	}
+	return &v3routepb.HeaderMatcher{
+		Name:                 key,
+		HeaderMatchSpecifier: &v3routepb.HeaderMatcher_ExactMatch{ExactMatch: value},
+	}
+}
+
+func parsePrincipalNames(principalNames []string) []*v3rbacpb.Principal {
+	var or []*v3rbacpb.Principal
+	for _, principalName := range principalNames {
+		newPrincipalName := &v3rbacpb.Principal{
+			Identifier: &v3rbacpb.Principal_Authenticated_{
+				Authenticated: &v3rbacpb.Principal_Authenticated{
+					PrincipalName: getStringMatcher(principalName),
+				},
+			}}
+		or = append(or, newPrincipalName)
+	}
+	return or
+}
+
+func parsePeer(source peer) *v3rbacpb.Principal {
+	var and []*v3rbacpb.Principal
+	if source.Principals != nil {
+		or := parsePrincipalNames(source.Principals)
+		if len(or) > 0 {
+			and = append(and, principalOr(or))
+		}
+	}
+	if len(and) > 0 {
+		return principalAnd(and)
+	}
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_Any{
+			Any: true,
+		},
+	}
+}
+
+func parsePaths(paths []string) []*v3rbacpb.Permission {
+	var or []*v3rbacpb.Permission
+	for _, path := range paths {
+		newPath := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_UrlPath{
+				UrlPath: &v3matcherpb.PathMatcher{
+					Rule: &v3matcherpb.PathMatcher_Path{Path: getStringMatcher(path)}}}}
+		or = append(or, newPath)
+	}
+	return or
+}
+
+func parseHeaderValues(key string, values []string) []*v3rbacpb.Permission {
+	var or []*v3rbacpb.Permission
+	for _, value := range values {
+		newHeader := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_Header{
+				Header: getHeaderMatcher(key, value)}}
+		or = append(or, newHeader)
+	}
+	return or
+}
+
+func parseHeaders(headers []header) ([]*v3rbacpb.Permission, error) {
+	var and []*v3rbacpb.Permission
+	for i, header := range headers {
+		if header.Key == "" {
+			return nil, fmt.Errorf("\"headers\" %d: \"key\" is not present", i)
+		}
+		// TODO(ashithasantosh): Return error for unsupported headers- "hop-by-hop",
+		// pseudo headers, "Host" header and headers prefixed with "grpc-".
+		if header.Values == nil {
+			return nil, fmt.Errorf("\"headers\" %d: \"values\" is not present", i)
```

Comment:
Use backticks instead:
```go
return nil, fmt.Errorf(`"headers" %d: "values" is not present`, i)
```

Should this error print the `header.Key` instead of the int?

Reply:
We use the index here, so that the error log points us to the part of the policy that was invalid. Since it is possible that the same header name could be used across multiple rules.